cpp/include/cudf/io/detail/batched_memset.hpp

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/resource_ref.hpp>
+
+/**
+ * @brief A helper function that takes in a vector of device spans and memsets them to the
+ * value provided using batches sent to the GPU.
+ *
+ * @param bufs Vector with device spans of data
+ * @param value Value to memset all device spans to
+ * @param _stream Stream used for device memory operations and kernel launches
+ * @param _mr Device memory resource used to allocate the returned column's device memory
+ *
+ * @return The data in device spans all set to value
+ */
+
+namespace CUDF_EXPORT cudf {
+namespace io::detail {
+
+void batched_memset(std::vector<cudf::device_span<uint64_t>>& bufs,
```

Comment:
Why are the buffers typed as `uint64_t`? I thought this would be a generic function with support for any fixed-width types. Do we not need the ability to set buffers of `std::byte` to a value?

It seems like we're being too specific to the use case in Parquet if the only type we care about is `uint64_t`. If that's the case, should we move this to `src/io/parquet` or something like that?

Reply:
I think this is okay as one can always initialize a std::byte buffer by providing a 8 byte value for it (say: -1UL) to initialize all bytes as 0xFF. I am a bit concerned about __int128_t buffers if we aren’t initializing with trivial 0s or -1ULs

Reply:
This doesn’t work for std::byte buffers unless their size is a multiple of 8, right?

Reply:
It relies on buffer allocation being 8 byte aligned as per the comment.

Reply:
Can we design this as a templated function that supports arbitrary fixed width types, or narrow the scope so this is more clearly a utility for only one specific use case in Parquet? Currently it seems like it is positioned as a generic IO utility but it is not generic.

Reply:
I think the goal here is to cast all buffers to a common type for us to use the cub function to memset all of them at once. Casting to uint64 does seem okay here as long as we can guarantee 8 byte aligned allocations (?) and are writing trivial all zeros or -1s for int128 and other types (currently true for Parquet output column buffers at least) and any other value for all other types (not sure if we do this in any reader for now).

Reply:
Alignment and length are different. Yes, 8 byte alignment would be required. But how would this work for a 4-byte allocation? Even if it is aligned (the starting pointer is on an 8 byte boundary), you can’t write a uint64_t because it would write four bytes out of bounds.

Reply:
For working with variable-length buffers like this, using std::byte as the common type seems like the only solution. Reinterpreting other allocations as uint64_t* is not safe.

Reply:
> Alignment and length are different. Yes, 8 byte alignment would be required. But how would this work for a 4-byte allocation? Even if it is aligned (the starting address is on an 8 byte boundary), you can’t write a uint64_t because it would write four bytes out of bounds.

Definitely. By saying 8 byte aligned alloc I meant as in the alloc size to be 8 byte aligned. 

>  // Using uint64_t as we are relying on the allocation padding for the buffers to be large enough
>  // that we can always write a multiple of 8 byte words

From this comment on  reader_impl_preprocess.cu it looks like we are relying on the assumption that rmm is padding the buffer allocations to be have 8 byte aligned allocated sizes to not go out of bounds. (I could only find this info https://github.com/rapidsai/rmm/pull/1278 where it looks like it has been removed) so it should be done explicitly in column_buffers.hpp functions.

>For working with variable-length buffers like this, using std::byte as the common type seems like the only solution. Reinterpreting other allocations as uint64_t* is not safe (UB).

Certainly if we can ensure that the value to memset would always fit in 1 byte even if we are memsetting a int4 or int8 buffer - the same concern I had with `int128_t` with the current design.

Reply:
> Though, with `std::byte` I am concerned we may lose any performance advantage we are currently seeing.

I discussed this with @abellina and @sdrp713. CUB decides [how many items-per-thread to use](https://github.com/NVIDIA/cccl/blob/02378eb73f214646669f98f661a6670e546a3a35/cub/cub/device/dispatch/dispatch_batch_memcpy.cuh#L125-L127) based on some heuristics. This should be efficient, and if it's not, we can file an issue with CUB.